### PR TITLE
3.4 sourceids targetids

### PR DESCRIPTION
**Pull request by mneedham**
This PR adds functionality discussed on the Neo4j community site - https://community.neo4j.com/t/algo-similarity-jaccard-stream-takes-more-than-3-minutes/4586/14

Things still to do:

- [x] Implement for Cosine Similarity
- [x] Implement for Pearson Similarity
- [x] Implement for Cosine Similarity
- [x] Implement for Overlap Similarity

### CODE REVIEW

**Comment by mneedham** [comment]
Simple benchmark with a dummy compute doesn't show much difference in the time taken to compute all pairs:

Benchmark                                                     (concurrency)  Mode   Cnt         Score         Error  Units
SimilarityStreamGeneratorBenchmark.allPairs                               1    ss  1000    138249.941 ±   16260.097  ns/op
SimilarityStreamGeneratorBenchmark.allPairs                               2    ss  1000  22361607.769 ± 2862052.577  ns/op
SimilarityStreamGeneratorBenchmark.allPairs                               8    ss  1000  19253922.198 ± 2185645.041  ns/op
SimilarityStreamGeneratorBenchmark.allPairsBlankSourceTarget              1    ss  1000     82898.555 ±    2002.081  ns/op
SimilarityStreamGeneratorBenchmark.allPairsBlankSourceTarget              2    ss  1000  18480648.936 ± 2548120.298  ns/op
SimilarityStreamGeneratorBenchmark.allPairsBlankSourceTarget              8    ss  1000  20026972.757 ± 2267054.050  ns/op

**Comment by mneedham** [comment]
[graph-algorithms-algo-3.4.12.3.zip](https://github.com/neo4j-contrib/neo4j-graph-algorithms/files/2879346/graph-algorithms-algo-3.4.12.3.zip)

File: algo/src/main/java/org/neo4j/graphalgo/similarity/SimilarityInput.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    static long[] extractInputIds(SimilarityInput[] inputs) {
+        return Arrays.stream(inputs).mapToLong(SimilarityInput::getId).toArray();
+    }
+
+    static int[] indexesFor(long[] inputIds, ProcedureConfiguration configuration, String key) {
+        List<Long> sourceIds = configuration.get(key, Collections.emptyList());
+        try {
+            return indexes(inputIds, sourceIds);
+        } catch(IllegalArgumentException exception) {
+            String message = String.format("%s: %s", String.format("Missing node id in '%s' list ", key), exception.getMessage());
```
[the remark this answers is not in the document]
Comment (mneedham) [reply algo/src/main/java/org/neo4j/graphalgo/similarity/SimilarityInput.java]: Good idea. Have done that

File: algo/src/main/java/org/neo4j/graphalgo/similarity/SimilarityInput.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+                indexes[indexesFound] = index;
+                indexesFound++;
+            }
+        }
+
+        return Arrays.copyOfRange(indexes, 0, indexesFound);
+    }
+
+    static long[] extractInputIds(SimilarityInput[] inputs) {
+        return Arrays.stream(inputs).mapToLong(SimilarityInput::getId).toArray();
```
[the remark this answers is not in the document]
Comment (mneedham) [reply algo/src/main/java/org/neo4j/graphalgo/similarity/SimilarityInput.java]: I think the sorting is handled by these bits of code:

https://github.com/neo4j-contrib/neo4j-graph-algorithms/blob/3.4/algo/src/main/java/org/neo4j/graphalgo/similarity/SimilarityProc.java#L275

and 

https://github.com/neo4j-contrib/neo4j-graph-algorithms/blob/3.4/algo/src/main/java/org/neo4j/graphalgo/similarity/SimilarityProc.java#L354

and 

https://github.com/neo4j-contrib/neo4j-graph-algorithms/blob/3.4/algo/src/main/java/org/neo4j/graphalgo/similarity/SimilarityProc.java#L306

I think the streams stuff preserves the order of those lists?

File: algo/src/main/java/org/neo4j/graphalgo/similarity/SimilarityInput.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                throw new IllegalArgumentException(String.format("Node id [%d] does not exist in node ids list", idToFind));
+            }
+
+            if(index >= 0) {
+                indexes[indexesFound] = index;
+                indexesFound++;
+            }
+        }
+
+        return Arrays.copyOfRange(indexes, 0, indexesFound);
```
[the remark this answers is not in the document]
Comment (mneedham) [reply algo/src/main/java/org/neo4j/graphalgo/similarity/SimilarityInput.java]: You're right. I was previously going to just return a list of ids that were there, but changed my mind half way.

File: algo/src/main/java/org/neo4j/graphalgo/similarity/SimilarityInput.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.List;
+
+public interface SimilarityInput {
+    long getId();
+
+    static int[] indexes(long[] inputIds, List<Long> idsToFind) {
+        int[] indexes = new int[idsToFind.size()];
+
+        int indexesFound = 0;
+        for (int i = 0; i < idsToFind.size(); i++) {
```
[the remark this answers is not in the document]
Comment (mneedham) [reply algo/src/main/java/org/neo4j/graphalgo/similarity/SimilarityInput.java]: Done

File: algo/src/main/java/org/neo4j/graphalgo/similarity/SimilarityInput.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jexp) [reply algo/src/main/java/org/neo4j/graphalgo/similarity/SimilarityInput.java]: Yes order is preserved.

**Pull request merged**